delphi-frontend/src/test/java/au/com/integradev/delphi/symbol/resolve/TypeInferrerTest.java

```diff
@@ -0,0 +1,334 @@
+  }
+
+  private static Type intrinsicType(IntrinsicType intrinsic) {
+    return TYPE_FACTORY.getIntrinsic(intrinsic);
+  }
+
+  private static ExpressionNode arrayConstructor(ExpressionNode... elements) {
+    var arrayConstructor = makeSpy(new ArrayConstructorNodeImpl(DelphiLexer.TkArrayConstructor));
+    arrayConstructor.addChild(commonDelphiNode(DelphiLexer.SQUARE_BRACKET_LEFT, "["));
+    for (ExpressionNode element : elements) {
+      arrayConstructor.addChild(element);
+    }
+    arrayConstructor.addChild(commonDelphiNode(DelphiLexer.SQUARE_BRACKET_RIGHT, "]"));
+
+    var primaryExpression = makeSpy(new PrimaryExpressionNodeImpl(DelphiLexer.TkPrimaryExpression));
+    primaryExpression.addChild(arrayConstructor);
+
+    return primaryExpression;
+  }
+
+  private static ExpressionNode integerLiteral(String image) {
+    var integerLiteralToken = new CommonToken(DelphiLexer.TkIntNumber, image);
+    var integerLiteral = makeSpy(new IntegerLiteralNodeImpl(integerLiteralToken));
+
+    var expression = makeSpy(new PrimaryExpressionNodeImpl(DelphiLexer.TkPrimaryExpression));
+    expression.addChild(integerLiteral);
+
+    return expression;
+  }
+
+  private static ExpressionNode procedural(Type returnType) {
+    return genericExpression(TYPE_FACTORY.routine(Collections.emptyList(), returnType));
+  }
+
+  private static ExpressionNode genericExpression(IntrinsicType type) {
+    return genericExpression(intrinsicType(type));
+  }
+
+  private static ExpressionNode genericExpression(Type type) {
+    var primaryExpression = makeSpy(new PrimaryExpressionNodeImpl(DelphiLexer.TkPrimaryExpression));
+    doReturn(type).when(primaryExpression).getType();
+    doReturn("<expression of type " + type.getImage() + ">").when(primaryExpression).getImage();
+    return primaryExpression;
+  }
+
+  private static ExpressionNode parenthesize(ExpressionNode expression) {
+    var parenthesized = new ParenthesizedExpressionNodeImpl(DelphiLexer.TkNestedExpression);
+    parenthesized.addChild(commonDelphiNode(DelphiLexer.PAREN_BRACKET_LEFT, "("));
+    parenthesized.addChild(expression);
+    parenthesized.addChild(commonDelphiNode(DelphiLexer.PAREN_BRACKET_RIGHT, ")"));
+    return parenthesized;
+  }
+
+  @SuppressWarnings("ObjectToString")
+  private static <T extends DelphiNodeImpl> T makeSpy(T spied) {
+    final T result = spy(spied);
+
+    doReturn(TYPE_FACTORY).when(result).getTypeFactory();
+    // This is just to improve the display string in parameterized tests
+    doAnswer(invocation -> result.getImage()).when(result).toString();
+
+    return result;
+  }
+
+  private static CommonDelphiNode commonDelphiNode(int type, String text) {
+    return new CommonDelphiNodeImpl(new CommonToken(type, text));
+  }
+
+  private static void assertInferred(ExpressionNode expression, Type type) {
+    doAssertInferred(expression, type);
+    if (expression != null) {
+      doAssertInferred(parenthesize(expression), type);
```

Comment:
What's the reasoning behind this?

Reply:
It ensures we're skipping parentheses in `TypeInferrer`.